### ISSUE TITLE
Pregen world prevents sluice from working

### ISSUE THREAD

**Issue by Nagapito** [Haggle1996/RevolutionPack#462]
Pregenarating the world, at least with the PregenSpawn command  does not generate the HardOre slice data for the chunk and the sluice will think there are no more ores available on the chunk and produce nothing.  

My pregenerated world has the HardOreData generated for the chunks around spawn that were generated normally. The chunks that were pegenerated have no HardOreData, even with ores in it.

Time to start a new world...

**Comment by Haggle1996** [comment]
@Draco18s, fyi. We tell players to use AdminCommandsToolbox if they want to pregenerate a world, though the pack doesn't require it.

Thanks, @Nagapito, for reporting this is important and may solve some of the sluice problems. Just to clarify, this world was generated using the latest version of REV|2?

**Comment by Nagapito** [comment]
it might actually not be related with the pregeneration.  

So, started a new world, wondered around until I find a nice place to set home and decided to cheat in the items and test it before I waste more time. Glad I did it. 

Sluice not working again. I googled a little and actually found your closed issue about the sluice and flowers not working. So, I tried the bone meal and no flowers. This, while seeing 4 iron ore blocks on a hole 7 blocks away from me. Counted the iron blocks with COFH, eighty something!

Seems the previous issue is still not totally fixed and sometimes the data is not generated...
Could it be that when the world generation is stressed, it doesnt have time to generate the data? I was flying in creative just looking around for a nice place to start my game and generating chunks like a mad man...  

if @Draco18s needs, I am available to test with a dev version. I am a dev to, logs dont scare me :)

**Comment by Haggle1996** [comment]
Something just occurred to me, @Nagapito. Did you generate with Vanilla world type instead of RTG or BoP?

**Comment by Draco18s** [comment]
BTW, what version of my mod are we running here?

**Comment by Haggle1996** [comment]
If he's on the latest version, it's 1.7.10-HarderOres-14.25.1e

**Comment by Draco18s** [comment]
That's the question: if he's not latest then the problem is that he needs to update.  If he is...hoo boy.

**Comment by Haggle1996** [comment]
You're not going to like this, but I tested it on 14.25.1e with the same seed, both pregenerated and not pregenerated. The good news is that this is very easy to replicate. :) 

On a pregenerated world, all flower/ore data is missing. It appears as though it's actually deleted, as even the spawn chunks are stripped of their ore data.

On the same seed, but flying manually to the same several spots, the sluices produce ore and flowers spawn with bonemeal, even in the spawn chunks. 

In the interim, I'm going to remove AdminCommandsToolbox and recommend players do not pregenerate.

**Comment by Haggle1996** [comment]
Just some insight into how AdminCommandsToolbox works: it generates 1000 chunks before actually saving chunks to disk. I don't know if possibly there's some array or similar that's overflowing (or what I would even look for in the log).

**Comment by Draco18s** [comment]
That's interesting.  I don't know what AdminCommandsToolbox does either (I've never used it) but it does sound like something it does skips over an event required by my mod to save the data.  My guess is that _all_ mods that have per-chunk data like this would be effected.

Thing is, my mod is supposed to detect the missing data and do a re-scan.

**Comment by Toksyuryel** [comment]
I believe Immersive Engineering adds per-chunk data for its excavator veins, so it would be worth checking to see if those happened to spawn in the pregenerated world.

**Comment by Nagapito** [comment]
Sorry for the delay, I went to bed.  

Latest version of the pack with the latest version of the mod. I even checked Curse to see if there was a new version.  

My first world with this issue was pre-generated and played on a client. This second world, is running on a server and chunks were generated by me flying crazy fast and waiting for them to show up, so I was making the server strugle with world generation.  

While ACT does not save the chunks immidiatle, so doesnt Minecraft. It would take a couple minutes for MC to actually write stuff on disk. I noticed this when forcing worldgen and waiting for it appear on the files so I could check them with NBTExplorer. And, to be honest, its good MC does not write to the disk every single time you break a block or a cow moves... If it did it, I would need a new SSD every month!  

I just searched for my chunk on the save files and HardOreData NBT only has one entry called 'version' with the value 2. Same for all the chunks around it. This are chunks that actually contain HardOres, since I can see them.  

And this might also be affecting other mods, since WildlifeTreeData nbt data also only contains one entry with the name version and value 2! But this might also mean nothing since it also only has this entry in chunks that HardOreData actually has proper values.  

If you want, make a a special version of the mod that logs everything about the missing data and rescan and see if we can figure out why rescan is not working.

**Comment by Nagapito** [comment]
My skills reading java code are not the best, I am a C# developer, so dont get too mad if I incorrectly assume something. 

On your OreDataHooks, your ReadData function will only fail and force a recount of ores it doesnt find the NBT tag HardOreData. But the tag exists with the version=2 child... So, it assume the chunk has the information generated and tries to read the details, but since there are no details, it never adds the hash with the data to your graphs storage. 

Later, the getOreData function, will try to access this hash in the graph, but since it does not exist, the function returns 0 instead of starting a recount when the hash is empty.  

Now, the part that I am not sure. The chunkLoad and chunkSave are events. True nature of events mean they are async. If they are async, there is nothing that can prevent the chunkSave event to be fired right after the chunkLoad, before the chunkLoad event as time to finish. So, you would be saving an empty NBT data...Under heavy load and low memory available, it makes sense the engine tries to save and unload chunks the fastest possible, to release memory.

My suggestion:
save function should not always create the NBT tag. It should only create the tag when it finds the first ore to save.
Load function should check if there are actually slices inside the NBT tag. If not, recalc.
getOreData should recalc if the map is null, since it means that the map was never loaded!

Or I could be totally wrong and out of mt league :)

**Comment by Draco18s** [comment]
We can certainly try doing that, see if it helps.
The whole chunk save/load/unload process is woefully annoying.  The worst part is that the Unload event (to remove data from the hashmap) and the Save event (to write the data) have virtually no relation to each other (e.g. chunks can get saved without being unloaded).  There was a point during my original coding--before anything was released--where I'd get an event sequence like this (for a single chunk):
Save
Unload
Save

So it would literally save good data, remove the data from the map, then be asked to save that data _again_ and not have it.

**Comment by Draco18s** [comment]
Alright, give this a go.
https://dl.dropboxusercontent.com/u/7950499/1.7.10-HarderOres-14.25.2b.jar

You may need the updated Lib, but I don't think so (some API alterations were made for Harder Wildlife last week, although that update isn't complete yet, there's still a persistent issue that needs to be tracked down).  But here it is, if you do need it.
https://dl.dropboxusercontent.com/u/7950499/1.7.10-HardLib-14.25.2b.jar

**Comment by Nagapito** [comment]
It actually makes sense.

Multiple saves can be called while its loaded and always call a last save after then unload to make sure that all changes between last save and the unload are saved and nothing is lost.

An easy solution for that is, on the unload, force a save. So, when it asks for the unload, you save the NBT with the latest valid data and unload the data from the map. The next save, will do nothing since there are is no data on the map and the chunk keeps the last saved data from the unload.

This is, assuming that if you dont save nothing to the NBT, what was previously there is not lost.
If its lost, then, you need a second map, where you store the unloaded data and on save, if you dont find it in the normal map, you search of the unloadedMap. If its on the unloadedMap, save that and remove it from the UnloadedMap.

Ok, testing the new version. Will report back in a few minutes

**Comment by Draco18s** [comment]
Can't save data during an Unload event: no access to the chunk NBT.  _That's_ why it's a problem!

**Comment by Nagapito** [comment]
I dont bring good news... Since the majority of the chunks didnt had the data generated, at load time its trying to generate data for all of them and taking a long time. So long, that the client entered that weird state where the world is still loaded and running but you are back on the main menu screen.

Restarted client and tried it again, but for some reason, its crashing while trying to track some entity.
Fresh backup, tried again and same issue... 
Crash log only from loading the world, http://paste.atlauncher.com/view/b6be0b15

This made me think of a couple of things, while I waited.
First, how MC reacts if you are scanning a chunk while its getting unloaded.You know, loading world, loads some chunks and unloads them right immediately. Might cause a lot of issues reading/saving the NBT

Second and more importantly. You might not need any of this. if I understood you are reading/writing to the chunk information that you only use in two scenarios, bonemealing and sluicing. Why use this heavy and prone to bugs mechanic? Why not calculate the mapping of the chunk at runtime when one of this tasks require it? 

Sluice starts, no mapping, scan the chunk. If there is a map, use the map. Chunk unloads, remove the mapping. Dont store it or rebuild it from stored data. Sluice scans a 3x3 area? big deal. Much better then a 8x8 or 10x10 for each player...And this will only happen once per loaded chunk on sluicing or bonemealing!

I think that removing the NBT dependency will speed up the loadings, remove the bugs and cause almost negligible load on clients or servers. Sometimes, is best to just recalc the data then store it for later use.

Unless I am missing something!

PS: Totally missed the detail of no NBT on unloading. That is... weird, specially if you can trigger a save after unload. This is the bad java conventions. Event should be called chunkUnloaded, so, no access to his data since its already unloaded. ChunkUnload to me its a ChunkUnloading, so you should still have access!

**Comment by Nagapito** [comment]
Also forgot to mention that, as you can notice in the logs, chunks are being scanned twice...

**Comment by Nagapito** [comment]
Also, reverted to the 'latest' oficial version, game loaded without entity tracking issues but only one rescan message on chunk 0,0

**Comment by Draco18s** [comment]
I can't calculate at use-time for two reasons:
1) Mining blocks shouldn't alter the data (the 'dust' had already leached info the soil, etc.)
1b) Silk Touch and _placing_ ore shouldn't alter the data!
2) Using the sluice depletes the data (because it is "free" ore, you shouldn't be able to set it atop a couple ore blocks and sluice it forever)

**Comment by Draco18s** [comment]
Alright, try this instead
https://dl.dropboxusercontent.com/u/7950499/1.7.10-HarderOres-14.25.2c.jar

It queues up rescan operations and does 1 per 5 ticks.

**Comment by Nagapito** [comment]
Rescan queue mechanic seems solid except for the part where it never fired :)

**Comment by Nagapito** [comment]
ohh, and I thought the sluice was depleting the ore blocks. At least that is what I read somewhere.
So, after a long time, if I tried to manually mine, the ore blocks would be empty. That is why I thought there was no need to rescan, since the data would be equal to the ore blocks in the chunk.

**Comment by Draco18s** [comment]
FFS. _Chews on the event_

> depleting the ore blocks

No no, it doesn't actually destroy any ore blocks anywhere.

**Comment by Draco18s** [comment]
Ok, the event works for me:
`[16:43:42] [Server thread/INFO] [STDOUT]: [com.draco18s.ores.OresEventHandler:worldTickEnd:341]: World Tick Event Called`
`[16:43:42] [Server thread/INFO] [HarderOres]: Chunk [0](-26,0,-4) is missing ore data, it will be rescanned.  Chunks way out on the edge of the world may not save and cause this message to repeat next launch; do not be alarmed.`
`[16:43:42] [Server thread/INFO] [HarderOres]: 488 chunks remain.`

**Comment by Nagapito** [comment]
I will give it another try and move around to load and unload chunks.

I only loaded the and stared at the console...
Also, running it on SSP, server takes even more time to stop/start. Shouldnt affect it, right?

**Comment by Nagapito** [comment]
Nope, nothing on the console.

Moved around, generated new chunks... TPS at 20.

Unless doRescan is actually called but doing nothing. Failing on the if statement...

**Comment by Draco18s** [comment]
Are you seeing INFO level notifications?

Because this is the code involved:

http://pastebin.com/wU6kQ119

**Comment by Nagapito** [comment]
You can confirm it yourself, http://paste.atlauncher.com/view/0a4c8d7f
And, yes, that is the code I am seeing when I decompile the jar that is being used in the client.
Also checked in-game mod list to confirm I really have the correct version, 14.25.2c

**Comment by Nagapito** [comment]
My decompiled code has a slight difference,
`if ((event.phase == TickEvent.Phase.END) &&  (event.world.func_72820_D() % 5L == 0L)) {`

But mod of 2 or 5 should still produce a 0...

**Comment by Draco18s** [comment]
The %2 was a change I made in dev to just get my machine to chew through entries faster (I was loading a flat world which _has_ no ore).

Try generating a new world.

**Comment by Nagapito** [comment]
Generated a new world and it worked \o/

Stop the world, loaded it again, and it was still working.

Load the previous world that was not working and it start working!!! 
Why? No clue :P

**Comment by Nagapito** [comment]
I still see the rescan being called twice to same chunks
And the stupid constant load/unload of chunks my MC is also screwing the generation. Was down at 5300 chunks from 5500, I am again at 5500 chunks

http://paste.atlauncher.com/view/e0200bb4

Seems like this number is going to take a lot of time until going down

**Comment by Nagapito** [comment]
And suddenly it stopped, claiming there were still 5709 items in the queue.
Its just me or is that size desynced?

Reloading the world started it again and size jumped to 6800

ps: stopped generating again after a while. moved around, generated new chunks... nothing

**Comment by Nagapito** [comment]
What happens if you try to generate  data on unloaded chunks?
Like, it queues 3k chunks to generate data but while its doing it, I move away 500 blocks. Those chunks are not loaded anymore... 

And every time I close and load a world, the queue size just gets bigger and bigger. Seems like the queue is not reseting when I reload a world.

Could it still have the queued chunks from previous reload and being unloaded causing the generator to stop?

At 10k now....

**Comment by Draco18s** [comment]
The code does not in any way attempt to figure out if the chunk is still in memory or not.  It was just a "make a queue" it's not optimized yet.

The counter on how many chunks are left is simply a counter of how many items are left in the queue, but due to the asynchronous nature of load/rescan, it is possible to add new chunks to the queue while it's still scanning.

Don't know why it would stop scanning with remaining chunks left. It does check that the chunk that's about to be rescanned is part of the world being ticked, but that shouldn't prevent the rescan from completing unless the chunk next in the queue is part of a world that is fully unloaded (e.g. inactive Mystcraft dimensions).

**Comment by Nagapito** [comment]
So, if for some reason, a chunk is momentarely loaded on another dimension and unloaded again, since that dimension is not ticking anymore, the whole system stops.

That could explain why my generation stops, since this packs has like 50 dimensions and most of them are constantly loaded (hate that feature in galacticraft...)

I also uploaded the new version to the server and he never did try generating the information... probably a similar issue?

**Comment by Draco18s** [comment]
Could be.
I'll pick this up tomorrow and do some more clever queue shuffling.

**Comment by Nagapito** [comment]
Just restarted the server and its now generating, even before I joined it.

The most amazing thing is it is generating for chunks everywhere!! Chunks like -50,-50 or 0,4 or, 20,55.
Not all this chunks so far apart should be loaded.... 
Something is really messy on how MC loads chunks

**Comment by Nagapito** [comment]
Suggestion: Add a config option to blacklist dimensions for harderOres or at least for generating the mapping. No point generating info on dimensions that will never have harder ores ore sluices like nether.

I just notice after I left the nether for the 1st time, queued chunks skyrockets to 100k and sent my TPS down to 12...

**Comment by Haggle1996** [comment]
@Draco18s any movement on this?

**Comment by Draco18s** [comment]
I did not get back to it "tomorrow." I try to devote my weekend time to fixing my mod bugs as my weekday hours are on my other project.  "Weekend hours" basically didn't happen this last weekend.

I haven't forgotten though.

**Comment by Haggle1996** [comment]
No worries -- I know you stay on top of stuff; I just wanted to make sure there wasn't another 'oh, forgot to publish that' moment. :D

**Comment by Draco18s** [comment]
![Whew](http://www.skype-emoticons.com/images/emoticon-00141-whew.png)!
Finally got around to making the queue system smarter.  It now has a separate queue per dimension (which can be blacklisted, blacklisted dimensions skip the entire ore data system, later re-enabling it will cause every chunk to be put into the queue when the world loads).  Dims -1 and 1 (Nether and End) are blacklisted by default.
@Nagapito 
https://dl.dropboxusercontent.com/u/7950499/1.7.10-HarderOres-14.25.2d.jar

**Comment by Haggle1996** [comment]
Oooh...new bits to test!

**Comment by Haggle1996** [comment]
I can't get any flowers to spawn using bonemeal with 25.2d

EDIT: And of course, the sluices don't work for same reason. On the plus side, chunks generate faster in the non-overworld. I assume that's related to your blacklist. :D

**Comment by Draco18s** [comment]
I got flowers in dev just fine.

**Comment by Nagapito** [comment]
Confirm, nothing is happening... 

Dont see any info on logs about generating the information, bonemeal does not work...  

Could you make a version that logs everything, that just spams my console! 
Like, 'Entered TickEnd event for dimensionID: x', before the if, calling doRescan if it passes the if....
Just log every step before a condition and after it, so we can understand if its entering the condition and if not, why?

I ask this because, with all the extra mods and extra load on the client that I have, your code might behave a little different then only with a few mods.

**Comment by Haggle1996** [comment]
Every time I load the same world, I get this exact message, but bonemeal and flowers do not work in the spawn zone, or out at a distance from spawn. I never get messages for any other chunks.

```
[14:16:04] [Server thread/INFO] [HarderOres/]: Chunk [1](0,0,0) is missing ore data, it will be rescanned.  Chunks way out on the edge of the world may not save and cause this message to repeat next launch; do not be alarmed.
[14:16:04] [Server thread/INFO] [HarderOres/]: 1 chunks remain.
```

**Comment by Draco18s** [comment]
That message is for dimension _1_ not the overworld.

**Comment by Nagapito** [comment]
Since sometimes the generation fails to run for some strange reason, I tried opening and closing the same world until it fired. After like 6-7 (lost count), still nothing!  

I also notice that every time I loaded the world, I had less and less free memory, last time was 40 Mb and game froze (Started with 1Gb free after finishing loading a world). I think this might be related with the qeue being loaded with all the chunks but on world stop, its not cleared.

Also, the key you use for the queue might be a bad key. You should use something like a value type. An object, will always be different every time you create a new one, unless the equals method is correctly implemented and the HashMap actually uses it. If not, two objects will never be equal. A string like worldID+x + ";" + Z is going to be unique for the same chunk and will properly do his work. 

Maybe that is what is causing memory going up. It kept adding the chunk when it already is in the queue

**Comment by Haggle1996** [comment]
Odd, because dimension 1 is in my blacklist.

**Comment by Nagapito** [comment]
I am struggling reading the documentation for the Array.binarySearch() but... should you be checking for a value greater then 0 instead of lower? 

Isn't a negative number when the item is not found? So, if not found, its not on the blacklist and should be processed?

Edit: And that would be why it is ignoring all Overworld and processing the blacklisted dimensions like it happened to Haggle?

**Comment by Draco18s** [comment]
Yeah, that's what quirked my eyebrow.
And @Nagapito that's the derp I just fixed :D I copy-pasted the check to a buncha places, and forgot to flip the comparison on the one method.

https://dl.dropboxusercontent.com/u/7950499/1.7.10-HarderOres-14.25.2e.jar

**Comment by Draco18s** [comment]
Re objects as keys: I correctly implement Equals and HashCode. :)

**Comment by Nagapito** [comment]
[25/6/2016 23:06:12 PM] [Server thread/INFO] [HarderOres]: World ID #0 ticking for rescan (-1)

And you still have the binaryCheck doing the wrong comparation in the doRescan

EDIT: Ohhh, and on the readData too!!!! And all the other places. That or my decompiler sucks :)

**Comment by Draco18s** [comment]
Oh ha.  I had it wrong in two places.  Thanks.  I am apparently not all here today -..-
https://dl.dropboxusercontent.com/u/7950499/1.7.10-HarderOres-14.25.2f.jar

**Comment by Nagapito** [comment]
Are you sure?

For example, readData. If the dimension is not one of blacklisted, it just exists. Same for all other uses with binarySearch? (getOreData, subOreData, clearData, cleanup,...

I will test this new version, but just seeing the code, I dont have high hopes

Edit: Also, chunkGen() in OresEventHandler

Edit 2: Confirmed, still not working

**Comment by Draco18s** [comment]
Grumble, grumble. I even double checked my math.  Grumble grumble.  Stupid brain being stupid again.  That's why I like "contains"!
https://dl.dropboxusercontent.com/u/7950499/1.7.10-HarderOres-14.25.2g.jar

**Comment by Nagapito** [comment]
Not that important but you still missed the check on chunkGen function. Not world breaking since it will re-generate later... 

Still loading the pack...

**Comment by Draco18s** [comment]
D'oh, thanks for the catch.

**Comment by Nagapito** [comment]
It.... seems... to... be.... WORKING!!

Also another small but, doRescan function, whe  you output the number of chunks to rescan, you output the toRescan queue size instead of the size of the actual queue for the dimension!
So, it keeps saying '1 chunk remain'.

had to wait for the other message that reports the queue size for each dimensions to see the real value, 473. 

going to let it run for a while to see if the number actually decreases or stops, since the previous versions, it kept growing and never stopped

**Comment by Nagapito** [comment]
MC world loading and saving is very very annoying, since it seems to not be synchronous....

I generated a new world, so I could have a clear readout of what was happening. And this is with the wrong check on chunkGen, so... no chunks got generated with the information!

And, it worked!!! it took some time, for some chunks the information is generated multiple times since MC just doesnt wait for you but eventually it got saved!!

So, I would say you just need to fix the check on world gen, fix the log message with the number of remaining chunks and maybe, tone down the logging spam. 

I think the message that runs every few minutes telling how many chunks need to be calculated is nice for admins, but should only show the dimensions that actually have something to calculate and ignore blacklisted ones or ones with an empty queue. This message, with only a summary of what is doing by dimension every minute should be better info and less spam

**Comment by Draco18s** [comment]
The debug info is just debug info.  I'll be removing most of it for release.

**Comment by Haggle1996** [comment]
@Draco18s, did this get packaged up?

**Comment by Draco18s** [comment]
Thanks for the ping; life occurred.

**Comment by Nagapito** [comment]
By the way, Dimension blacklist is awesome!!

I dont need the information anymore, since I long pass the need for sluice and bonemeal and the constant data generation of those pesky border chunks was annoying me! Not that it was killing the server but... why use 20% of my CPU on something that I dont need?

So, I blacklisted the Overworld too. Server is now at 2-3% CPU when idle :)

It might even be a good tip for server admins...

**Comment by Haggle1996** [comment]
If I had my druthers, I'd rather it have been a dimension whitelist, prefilled with only the overworld, but I'm okay adding all known dimensions in my pack to the blacklist (it's already done -- waiting on the fix). 

The issue, I imagine, will be with dynamically-created dimensions -- think Mystcraft or RFTools, also GalactiCraft's space stations. Which would mean Draco18s would need to add a `isBlacklistWhite=false`, so that the behavior of the list could be flipped between black and white, depending on the whims of the pack author.

But I'm happy to get what I'm getting. Maybe he'll add something like that in the future, but 1.7.10 is on it's last legs, and Draco18s and I both have only a limited amount of time to dedicate to our hobby. :)

**Comment by Draco18s** [comment]
The reason for the Blacklist is for dimensions like Mystcraft Ages or RF whatever-its-called.  Those are as arbitrary as you can get and I'd rather _catch_ those on the fly than have to have them added to a white list.

**Comment by Haggle1996** [comment]
I think we're saying the same thing. :D 

Regardless, thank you for all your hard work. I and my players greatly appreciate it.

**Comment by Draco18s** [comment]
sure thing!

**Comment by Haggle1996** [comment]
@Draco18s notice anything wrong with this picture:
![image](https://cloud.githubusercontent.com/assets/11808184/16477068/ce9a3e50-3e4b-11e6-9a49-13952432ae20.png)

:)

**Comment by Draco18s** [comment]
D'oh.
Fixing now.

Wildlife and Underground got reversed.  Should be visible correctly in a few minutes.  Thanks for catching it before I went on vacation (I will be out of town until Tuesday).

**Comment by Haggle1996** [comment]
Thank you both for testing this!

**Comment by Haggle1996** [comment]
@Draco18s 

Is there a way to remove this: 
![image](https://cloud.githubusercontent.com/assets/11808184/16496141/c98c699e-3ead-11e6-826a-6208ffd09e95.png)

**Comment by Haggle1996** [comment]
BTW, all of these worlds are in the blacklist, except for 0

**Comment by Nagapito** [comment]
You sure all of them are on the blacklist?

Blacklisted ones should say -2 in the amount of ores to rescan...
-1 is when there isnt even a queue with chunks to rescan and 0 or higher is the size of the queue when it exists.

And I see a couple of -1's... Unless that is what you are trying to say...

**Comment by Haggle1996** [comment]
I think I'm saying 'I don't want these messages spamming the console constantly when I release this'...I really don't want it while I'm working on other things, as what I need to see scrolls out of the debug window after a few seconds. :)

**Comment by Draco18s** [comment]
Oh, that's what I forgot to do!
Remember I said I was going to remove a lot of debug info? Yeah. 
Unfortunately this will have to wait until next week. Sorry!

**Comment by Nagapito** [comment]
Tip for the future, instead of logging them as INFO, log them as DEBUG and ask who is helping you to test the mod to lower the log level, since all launchers/clients only show INFO and above levels.

This way, you dont have to do 'log cleanups' on your code and if something weird happens in the future, you can always ask the bug reporter to lower the log level and you will have the extra information without needing a new version.

**Comment by Draco18s** [comment]
I'll try to do that, although I recall doing that once and the person who
was helping test couldn't see them.

**Comment by Nagapito** [comment]
I think they need to pass a parameter on the java console to lower the log level on Log4. By default its at INFO and on development ambient is at the lower setting, so you could see them.

**Comment by Draco18s** [comment]
Yeah, I tried to explain that and didn't have any luck.
